Doc for 'preDNAT' policy flag

neiljerram opened the pull request:
## Description
Draft documentation for the forthcoming 'preDNAT' policy flag.

## Todo
- [x] Review, revise and merge.
- [x] Write release note.

## Release Note
<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
Pre-DNAT Policy - a new flavor of Calico Policy that is enforced before any DNAT that a cluster node may do (for example kube-proxy).  Pre-DNAT Policy is useful for securing the perimeter of a cluster against incoming traffic, except for pinholes that are expressed in terms of particular IP addresses and/or ports that external clients are allowed to connect in to.  For more information please see http://docs.projectcalico.org/v2.4/getting-started/bare-metal/bare-metal#pre-dnat-policy.
```

bcreane commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +427,24 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+
+## Pre-DNAT policy
+
+Policy for host endpoints can be marked as 'preDNAT'.  This means that rules in
```
Very well written section. One suggestion, consider expanding DNAT acronym once (if it's not already defined anywhere else) - I know not everyone is familiar with the term.

neiljerram replied:
Thanks @bcreane, I'll look at doing that.

lxpollitt commented on master/reference/calicoctl/resources/policy.md:
```diff
@@ -46,21 +46,23 @@ spec:
+| preDNAT    | Indicates to apply the rules in this policy before any DNAT.                                                                                          | true, false     | boolean               | false   |
+
+The `doNotTrack` and `preDNAT` fields are meaningful only when applying policy to a
+[host endpoint]({{site.baseurl}}/{{page.version}}/reference/calicoctl/resources/hostendpoint).
```
I didn't realize that `doNotTrack` was limited to host endpoints today.  I guess (assuming that is true) then preDNAT would be ok to limit to host endpoints too, but it would be good to better understand the implementation cost of making these options supported for all endpoint types.  I'm not advocating expanding the scope of this PR, but perhaps we should have a follow-on PR for consideration for Calico 2.5?

neiljerram replied:
Yes, `doNotTrack` has so far been limited to host endpoints.  Also I'm pretty sure that the current _requirement_ for preDNAT only needs application to host endpoints.

I think it would be a doable but non-trivial addition to do preDNAT for workload endpoints too -
 so happy for considering this as a possible feature for 2.5.  (Think we need to
- make sure that relevant workload endpoint chains are in the mangle table
- program workload dispatch chains into the mangle table
- add rules into Calico's mangle PREROUTING chain to dispatch to the workload chains)

fasaxc replied:
Couple of thoughts:
- There's a dataplane perf impact for every untracked rule that we add so I think there needs to be a clear requirement for workload untracked policy before we pay the per-packet cost for the dispatch chains.
- It'd be hard to get good semantics for untracked policy for workloads.  Since the raw PREROUTING chain is before the routing table and before NAT, we don't know whether the packets _are_ workload packets at the point the rules are executed.  Also, marking a packet as untracked takes it out of NAT so it breaks things like kube-proxy.

neiljerram commented:
@fasaxc - Would appreciate your review of the new "When do host endpoint policies apply" section, in case you spot any mistakes.

@tmjd - Would appreciate your thoughts also, to see if you're happy that this meshes well with your https://github.com/projectcalico/calico/pull/892

tmjd commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +427,90 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+  > overall result as if there was no untracked policy at all, so in practice we
+  > can describe this as untracked policies not applying to this flow.
+
+- Pre-DNAT policies _do_ apply.
```
Does this mean if I have a preDNAT policy that accepts all traffic for a NodePort, if I have workload also listening on that same port that the traffic will be allowed, even if I don't have a policy allowing it to that workload?
If so, we might want a note about that.

fasaxc replied:
Hmm, that's an excellent point.  Does the pre-DNAT policy do a "final" accept that skips workload policy?  That could be very confusing!

neiljerram replied:
Pre-DNAT policy doesn't know if it's going to a workload, because it's pre-routing.

neiljerram replied:
But also that's a key part of its feature - that it works the same for kube-proxy forwarding to a local pod as to a remote pod.

neiljerram replied:
(OK, I misunderstood the point here - let's talk.)

neiljerram commented:
@tmjd @fasaxc @lxpollitt I've added a lot to this doc and would appreciate your review again.  In particular:
- I believe I've addressed Erik's concern above, by clarifying that pre-DNAT policy and workload policy both apply, for traffic that comes in a host endpoint and goes to a local workload
- I've added a completed worked example for a typical use case, and hope that that now demonstrates (a) that pre-DNAT policy will be effective in achieving that use case, and (b) that the policy configuration required of the admin is as simple as it could reasonably be.

Please do let me know what you think.

fasaxc commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,280 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+Traffic from outside is addressed to any node's IP address, on a known
+NodePort, and Kubernetes (kube-proxy) then DNATs that to the IP address of one
+of the pods that provides the corresponding Service, and the relevant port
+number on that port (which is usually different from the NodePort).
```
Typo: `port number on that port`

fasaxc commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,280 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+reflecting that it is designed for the policing of incoming traffic from
+outside the cluster:
+
+1. Only the ingress rules of a pre-DNAT policy are enforced.  If the policy has
```
Out of date: we now validate against egress policy being supplied.

fasaxc commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,280 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+   (Whereas normal host endpoint policy is skipped, for traffic going to a
+   local workload.)
+
+3. There is no 'default drop' semantic for pre-DNAT policy (as there is for
```
I think this only applies if the packet is heading to a pod/service.  How about breaking this into cases for "packet terminated by host" (where the remaining normal policy then applies including default drop) and "packet forwarded on to workload or remote workload" (where the packet goes on its way)

neiljerram replied:
As discussed, think this is actually OK as is.

fasaxc commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,280 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+For packets that arrive on a host interface and are destined for a local
+workload - i.e. a locally-hosted pod, container or VM:
+
+- Untracked policies technically do apply, but never have any net positive
```
Bit of a shame to start this section with a negative "technically yes, but no".  Can you pull the mainline useful cases to the top and push this one to the bottom (maybe even in a "Not recommended" section)?

neiljerram replied:
To stay consistent, I'll move all of the (effectively-)non-applying cases to the end of their sections.  I thought it was nice to have the really-applying ones in the order that they happen, though.

fasaxc commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,280 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+  destination workload's ingress policy in this case.
+
+For packets that arrive on a host interface and are destined for a local
+server in the host namespace:
```
Would it be better to say `server process`, just to avoid the reading parsing it as `server machine` (as I did and had to backtrack)

fasaxc commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,280 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+
+## Pre-DNAT policy: a worked example
+
+Imagine a Kubernetes cluster, that its administrator wants to secure as much as
```
This is a great example but a diagram would be a great help here, it was at the limit of my mental "stack" to follow!

It might help to point out that you're talking about a VIP that routes to several hosts for the load balancer; I didn't grok that at first; left me wondering why several hosts could have the same IP in the dest.

tmjd commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,284 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+  NodePorts - i.e. as well-known TCP port numbers that appear to be available
+  on any node in the cluster.
+
+- Most of those Services, however, should not be accessed via _any_ node, but
```
Before I read through the examples below I thought you were going to block nodes and pods from being able to access these NodePorts too (if the traffic was not destined for the LoadBalancer IP).  I'd like to see it mentioned that the traffic being blocked is external traffic only.

neiljerram replied:
Thanks; adding two more instances of "from outside the cluster", to clarify this.

caseydavenport commented:
@neiljerram in case you forgot, this PR needs a real release note soon

neiljerram commented:
@caseydavenport Release note filled in.

@tmjd Thanks for your review.  I may be able to work on your comments during the weekend, but otherwise it will be Monday.

lxpollitt commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,284 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+   but has no pre-DNAT policies that explicitly allow or deny a particular
+   incoming packet, that packet is allowed to continue on its way, and will
+   then be accepted or dropped according to workload policy - if it is going to
+   a local workload - or to normal host endpoint policy - if not.
```
Would this punctuation be slightly easier to parse?
"then be accepted or dropped according to workload policy (if it is going to
a local workload) or to normal host endpoint policy (if not)."

neiljerram replied:
Yes, thanks.

lxpollitt commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,284 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+  ingress:
+    - action: deny
+      source:
+        notNets: [<pod-cidr>, <cluster-internal-node-cidr>, ...]
```
This works / is correct, but I'm wondering if a different example using 2 policies to achieve this might be clearer instead:
- First policy is a normal (not preDNAT) policy that you would setup on a node in a k8s cluster if you are using hostEndpoints.  This is the "allow source <pord-cidr>...".  You would do this independent of whether you were trying to lock down nodePorts.
- Second policy is a preDNAT policy that locks down the nodePorts "deny dest <node-port-range>".

WDYT?

neiljerram replied:
- I don't believe the "second policy ... that locks down the nodePorts" matches the use case that the rest of the narrative here describes, namely to lock down _all_ external traffic and then open pinholes for particular node ports or load balancer ingress IPs.
- I could perhaps adjust your suggestion for that, but given where we are w.r.t. the release, and that I'm not confident that I completely understand your suggestion, I'd rather leave this as is for now and use a new PR for further refinements.

lxpollitt commented on master/getting-started/bare-metal/bare-metal.md:
```diff
@@ -427,3 +430,284 @@ an ingress rule allowing access *to* port 999 and an egress rule allowing
+workload policy, for packets going to a local workload.)
+
+For the example here, that means that the last pre-DNAT policy above does not
+accidentally expose workloads that happen to use the same `<node-port>`, or
```
Yes, the behavior Neil describes is what we want (and is hopefully what he's implemented too).

neiljerram replied:
@tmjd I think the way to think of this, in general, is that the pre-DNAT policy protects the cluster perimeter, and the workload policy protects the particular workload; and then I think it's obvious that there will be situations where both of those are wanted.

caseydavenport commented:
@neiljerram the v2.4 docs have been merged so we'll need to copy these changes over to that directory as well.

neiljerram commented:
@tmjd Many thanks for your review comments.  I've addressed them specifically above, except for the one about the diagram.  I think you're right that it could be cleaner without the second row of Nodes, but unfortunately it's not easy for me to change this now - because I prepared it using Google Drawing and exported as PNG but unfortunately forgot to edit as some editable format like SVG!  I still think this may be worth doing, but probably not now before 2.4 goes out.

neiljerram commented:
@lxpollitt Thanks for your comments too, which I've addressed individually above.
